src/fs/metadata_log.cc

```diff
@@ -100,18 +100,55 @@ void metadata_log::write_update(inode_info::file& file, inode_data_vec new_data_
     } else {
         cut_out_data_range(file, new_data_vec.data_range);
     }
-
     file.data.emplace(new_data_vec.data_range.beg, std::move(new_data_vec));
 }
 
 void metadata_log::cut_out_data_range(inode_info::file& file, file_range range) {
-    file.cut_out_data_range(range, [](inode_data_vec data_vec) {
-        (void)data_vec; // TODO: for compaction: update used inode_data_vec
+    file.cut_out_data_range(range, [&](const inode_data_vec& former, const inode_data_vec& left_remains, const inode_data_vec& right_remains) {
+        std::visit(overloaded {
+            [&](const inode_data_vec::in_mem_data&) {
+                _compacted_log_size -= ondisk_entry_size<ondisk_small_write_header>(former.data_range.size());
+            },
+            [&](const inode_data_vec::on_disk_data& disk_data) {
+                // TODO: Differentiate between large write and large write without mtime
+                _compacted_log_size -= (former.data_range.size() == _cluster_size ?
+                    ondisk_entry_size<ondisk_large_write>() :
+                    ondisk_entry_size<ondisk_medium_write>());
+                auto it = _data_cluster_contents_info_map.find(offset_to_cluster_id(disk_data.device_offset, _cluster_size));
+                assert(it != _data_cluster_contents_info_map.end());
+                it->second->cut_data(disk_data.device_offset, former.data_range, left_remains.data_range, right_remains.data_range);
+            },
+            [&](const inode_data_vec::hole_data&) {
+            },
+        }, former.data_location);
+        auto visit_remains = [&](const inode_data_vec& data_vec) {
+            if (!data_vec.data_range.is_empty()) {
+                std::visit(overloaded {
+                    [&](const inode_data_vec::in_mem_data&) {
+                        _compacted_log_size += ondisk_entry_size<ondisk_small_write_header>(data_vec.data_range.size());
+                    },
+                    [&](const inode_data_vec::on_disk_data& disk_data) {
+                        // Remaining data is smaller than original so it is not a large write anymore
+                        assert(data_vec.data_range.size() < _cluster_size);
+                        _compacted_log_size += ondisk_entry_size<ondisk_medium_write>();
+                    },
+                    [&](const inode_data_vec::hole_data&) {
+                    },
+                }, data_vec.data_location);
+            }
+        };
+        visit_remains(left_remains);
+        visit_remains(right_remains);
     });
 }
 
 inode_info& metadata_log::memory_only_create_inode(inode_t inode, bool is_directory, unix_metadata metadata) {
     assert(_inodes.count(inode) == 0);
+    _compacted_log_size += ondisk_entry_size<ondisk_create_inode>();
+    // During rewrite we'll start with truncating the file to its final size to remember trailing holes
+    if (!is_directory) {
+        _compacted_log_size += ondisk_entry_size<ondisk_truncate>();
```

Comment:
I'm not sure which placement for the comment is better, what do you think @psarna?
```suggestion
    if (!is_directory) {
        // During rewrite we'll start with truncating the file to its final size to remember trailing holes
        _compacted_log_size += ondisk_entry_size<ondisk_truncate>();
```